Revive macOS CI tests

nilsvu opened the pull request:
## Proposed changes

Revive CI tests on macOS. Run on PRs, but don't require the tests to pass. The tests are useful for catching missing link dependencies, since macOS / AppleClang is stricter about those than Linux, and it's annoying when they break macOS builds on develop.

- [x] Shared libs on macOS work since #2680
- [x] Installation of dependencies on macOS with Spack is easier since #3712
- [x] Depends on #3773 

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe commented on docs/Installation/Installation.md:
```diff
@@ -19,7 +19,7 @@ installation_on_clusters "Installation on clusters" page.
 * BLAS (e.g. [OpenBLAS](http://www.openblas.net))
 * [Blaze](https://bitbucket.org/blaze-lib/blaze/overview) v3.5
-* [Boost](http://www.boost.org/) 1.60.0 or later
+* [Boost](http://www.boost.org/) between 1.63.0 and 1.72.0
```
Please don't make unrelated changes in your PRs.  This is only correct for OSX (or generally systems with case-insensitive file systems).  Higher versions work fine in other cases.

nilsvu replied:
Sure, I dropped this commit for now and will move it to a separate PR.

nilsvu committed:
Revive macOS CI tests

nilsdeppe commented on .github/workflows/MacOs.yaml:
```diff
@@ -110,17 +109,19 @@ jobs:
-          rm ../charm-${CHARM_VERSION}.tar.gz
+          git clone https://github.com/UIUC-PPL/charm && cd charm
+          # This is the commit that fixed #2397
+          git checkout e7e9d4b5b27608b89ab09f021b6b5194267f74fe
```
I'm confused, why does this install an old version of Boost to avoid the issue, but then check out a random hash to avoid the boost issue... I feel like we should only check out the exact Charm++ version we support

nilsvu replied:
#2397 has nothing to do with boost. Some `.tbd` files appeared for some reason on GitHub's latest macOS VM and Charm couldn't handle those files, so our CI build broke. This commit https://github.com/UIUC-PPL/charm/commit/e7e9d4b5b27608b89ab09f021b6b5194267f74fe is the one that fixes the issue. It will be in charm v6.11, but they'll probably not have another bugfix release before that. So I figured it's better to checkout the particular commit than to checkout whatever the latest `master` is when the CI build runs. It's the closest thing to a bugfix release we can do I think.

kidder commented:
marking as in progress until MacOS build passes

nilsvu commented:
Yep, I'll wait with this until some of the recent build system PRs are in that make it easier to compile on macOS.

nilsvu committed:
Set `PYTHONPATH` in tests

Fix issues with missing Python modules on systems where they are
made available through the `PYTHONPATH`.

nilsvu commented:
This is ready for review now. The macOS job takes ~1h using caches from `develop`, see https://github.com/nilsvu/spectre/runs/4996097358?check_suite_focus=true.

nilsvu commented:
@wthrowe or @kidder could you take a look at this one?

kidder commented on .github/workflows/Tests.yaml:
```diff
@@ -602,6 +602,158 @@ ${{ env.CACHE_KEY_SUFFIX }}"
+      - name: Build tests
+        working-directory: build
+        run: |
+          make -j4 RunTests
```
I believe that only two are available so `-j2`

nilsvu replied:
squashed this in and rebased

kidder commented:
@wthrowe you have a old requested change, so please look at this

wthrowe commented on cmake/AddInputFileTests.cmake:
```diff
@@ -5,6 +5,15 @@ spectre_define_test_timeout_factor_option(INPUT_FILE "input file")
 find_package(Python REQUIRED)
 
+# Environment variables for test
+set(_TEST_ENV_VARS "")
```
What's the leading underscore about?

nilsvu replied:
This is how it's named in `SpectreSetupPythonPackage.cmake` 🤷‍♂️ Some people prefix local variables with leading underscores in CMake.
